lib/helpers/query/castUpdate.js

```diff
@@ -196,7 +196,12 @@ function walkUpdatePath(schema, obj, op, options, context, filter, pref) {
       }
     }
 
-    if (schema.discriminatorMapping != null && key === schema.options.discriminatorKey && !options.overwriteDiscriminatorKey) {
+    if (
+      schema.discriminatorMapping != null &&
+      key === schema.options.discriminatorKey &&
```

Comment:
Let's make this `(prefix ? prefix + '.' + key : key) === schema.options.discriminatorKey &&`

Although we don't have test coverage for nested discriminator keys, it seems to work at first glance. For example, the following script works as expected.

```javascript
'use strict';

const mongoose = require('mongoose');
mongoose.set('debug', true);

const { Schema } = mongoose;

run().catch(err => console.err(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test');

  const schema = new Schema({ name: String, meta: { test: String } }, { discriminatorKey: 'meta.kind' });
  const Test = mongoose.model('Test', schema);
  const Disc = Test.discriminator('Disc', Schema({ test: String }));

  const disc = new Disc({ name: 'test' });
  await disc.save();

  console.log(await Disc.findById(disc));
}
```

Given that, it would be great if this check didn't break on if discriminator key has a `.`.

Reply:
👍 I'll work on this